File: ambeth/Ambeth.Merge.Bytecode/ambeth/bytecode/visitor/ObjRefVisitor.cs
using De.Osthus.Ambeth.Bytecode.Behavior;
using De.Osthus.Ambeth.CompositeId;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Metadata;
using De.Osthus.Ambeth.Mixin;
using De.Osthus.Ambeth.Typeinfo;
using De.Osthus.Ambeth.Util;
using System;
using System.Reflection;
using System.Reflection.Emit;
using System.Text;

namespace De.Osthus.Ambeth.Bytecode.Visitor
{
    public class ObjRefVisitor : ClassVisitor
    {
        public static readonly Type templateType = typeof(ObjRefMixin);

	    protected static readonly String templatePropertyName = templateType.Name;

        private static readonly ConstructorInstance c_stringBuilder = new ConstructorInstance(typeof(StringBuilder).GetConstructor(Type.EmptyTypes));

        private static readonly PropertyInstance template_p_realType = PropertyInstance.FindByTemplate(typeof(IObjRef), "RealType", typeof(Type), false);

        private static readonly PropertyInstance template_p_idIndex = PropertyInstance.FindByTemplate(typeof(IObjRef), "IdNameIndex", typeof(sbyte), false);

        private static readonly PropertyInstance template_p_id = PropertyInstance.FindByTemplate(typeof(IObjRef), "Id", typeof(Object), false);

        private static readonly PropertyInstance template_p_version = PropertyInstance.FindByTemplate(typeof(IObjRef), "Version", typeof(Object), false);

       	private static readonly MethodInstance template_m_equals = new MethodInstance(null, typeof(Object), typeof(bool), "Equals", typeof(Object));

	    private static readonly MethodInstance template_m_hashCode = new MethodInstance(null, typeof(Object), typeof(int), "GetHashCode");

        private static readonly MethodInstance template_m_toString = new MethodInstance(null, typeof(Object), typeof(String), "ToString");

        private static readonly MethodInstance template_m_toStringSb = new MethodInstance(null, typeof(IPrintable), typeof(void), "ToString", typeof(StringBuilder));

        private static readonly MethodInstance m_objRef_equals = new MethodInstance(null, templateType, typeof(bool), "ObjRefEquals", typeof(IObjRef), typeof(Object));

	    private static readonly MethodInstance m_objRef_hashCode = new MethodInstance(null, templateType, typeof(int), "ObjRefHashCode", typeof(IObjRef));

	    private static readonly MethodInstance m_objRef_toStringSb = new MethodInstance(null, templateType, typeof(void), "ObjRefToString", typeof(IObjRef),
			typeof(StringBuilder));

        public static PropertyInstance GetObjRefTemplatePI(IClassVisitor cv)
        {
            Object bean = State.BeanContext.GetService(templateType);
            PropertyInstance pi = State.GetProperty(templatePropertyName, NewType.GetType(bean.GetType()));
            if (pi != null)
            {
                return pi;
            }
            return cv.ImplementAssignedReadonlyProperty(templatePropertyName, bean);
        }

        public static PropertyInstance GetConversionHelperPI(IClassVisitor cv)
	    {
		    Object bean = State.BeanContext.GetService<IConversionHelper>();
		    PropertyInstance pi = State.GetProperty("ConversionHelper", NewType.GetType(bean.GetType()));
		    if (pi != null)
		    {
			    return pi;
		    }
		    return cv.ImplementAssignedReadonlyProperty("ConversionHelper", bean);
	    }

        protected readonly IEntityMetaData metaData;

        protected readonly int idIndex;

        public ObjRefVisitor(IClassVisitor cv, IEntityMetaData metaData, int idIndex)
            : base(new InterfaceAdder(cv, typeof(IObjRef)))
        {
            this.metaData = metaData;
            this.idIndex = idIndex;
        }

        public override void VisitEnd()
        {
            ImplementRealType();
            ImplementIdIndex();
            PropertyInstance p_id = ImplementId();
            PropertyInstance p_version = ImplementVersion();
            ImplementDefaultConstructor();
            ImplementIdVersionConstructor(p_id, p_version);
            ImplementToString();
            ImplementEquals();
            ImplementHashCode();

            base.VisitEnd();
        }

        protected void ImplementDefaultConstructor()
        {
            IMethodVisitor mv = VisitMethod(ConstructorInstance.defaultConstructor);
            mv.LoadThis();
            mv.InvokeSuperOfCurrentMethod();
            mv.ReturnValue();
            mv.EndMethod();
        }

        protected void ImplementIdVersionConstructor(PropertyInstance p_id, PropertyInstance p_version)
        {
            ConstructorInfo superConstructor = State.CurrentType.GetConstructor(Type.EmptyTypes);
            ConstructorInstance ci_super = new ConstructorInstance(superConstructor);

            IMethodVisitor mv = VisitMethod(new ConstructorInstance(MethodAttributes.Public, typeof(Object), typeof(Object)));
            mv.LoadThis();
            mv.InvokeConstructor(ci_super);
            mv.CallThisSetter(p_id, delegate(IMethodVisitor mg)
                {
                    mg.LoadArg(0);
                });
            mv.CallThisSetter(p_version, delegate(IMethodVisitor mg)
                {
                    mg.LoadArg(1);
                });
            mv.ReturnValue();
            mv.EndMethod();
        }

        protected void ImplementRealType()
        {
            ImplementProperty(template_p_realType, delegate(IMethodVisitor mg)
                {
                    mg.Push(metaData.EntityType);
                    mg.ReturnValue();
                }, delegate(IMethodVisitor mg)
                {
                    mg.ThrowException(typeof(NotSupportedException), "Property is read-only");
                    mg.ReturnValue();
                });
        }

        protected void ImplementIdIndex()
        {
            ImplementProperty(template_p_idIndex, delegate(IMethodVisitor mg)
                {
                    mg.Push(idIndex);
                    mg.ReturnValue();
                }, delegate(IMethodVisitor mg)
                {
                    mg.ThrowException(typeof(NotSupportedException), "Property is read-only");
                    mg.ReturnValue();
                });
        }

        protected PropertyInstance ImplementPotentialPrimitive(PropertyInstance property, Member member)
        {
            if (member == null)
            {
                return ImplementProperty(property, delegate(IMethodVisitor mg)
                    {
                        mg.PushNull();
                        mg.ReturnValue();
                    }, delegate(IMethodVisitor mg)
                    {
                        Label l_isNull = mg.NewLabel();

                        mg.LoadArg(0);
                        mg.IfNull(l_isNull);
                        mg.ThrowException(typeof(NotSupportedException), "Property is read-only");
                        mg.Mark(l_isNull);
                        mg.ReturnValue();
                    });
            }
            PropertyInstance p_conversionHelper = GetConversionHelperPI(this);

		    MethodInstance m_convertValueToType = new MethodInstance(ReflectUtil.GetDeclaredMethod(false, typeof(IConversionHelper), typeof(Object),
				"ConvertValueToType", typeof(Type), typeof(Object)));

            Type type = member.RealType;
            FieldInstance field = ImplementField(new FieldInstance(FieldAttributes.Private, "f_" + property.Name, type));
            return ImplementProperty(property, delegate(IMethodVisitor mg)
                {
                    if (member.RealType.IsValueType)
                    {
                        Label l_isNull = mg.NewLabel();
                        LocalVariableInfo loc_value = mg.NewLocal(field.Type);

                        mg.GetThisField(field);
                        mg.StoreLocal(loc_value);

                        mg.LoadLocal(loc_value);
                        mg.IfZCmp(field.Type, CompareOperator.EQ, l_isNull);

                        mg.LoadLocal(loc_value);
                        mg.Box(type);
                        mg.ReturnValue();

                        mg.Mark(l_isNull);
                        mg.PushNull();
                        mg.ReturnValue();
                    }
                    else
                    {
                        mg.GetThisField(field);
                        mg.ReturnValue();
                    }
                }, delegate(IMethodVisitor mg)
                {
                    mg.PutThisField(field, delegate(IMethodVisitor mg2)
                        {
                            Label l_isNull = mg2.NewLabel();
                            Label l_finish = mg2.NewLabel();

                            mg2.LoadArg(0);
                            mg2.IfNull(l_isNull);

                            mg.CallThisGetter(p_conversionHelper);
                            mg.Push(type);
                            mg.LoadArg(0);
                            mg.InvokeVirtual(m_convertValueToType);

                            mg2.Unbox(type);
                            mg2.GoTo(l_finish);

                            mg2.Mark(l_isNull);
                            mg2.PushNullOrZero(field.Type);

                            mg2.Mark(l_finish);
                        });
                    mg.ReturnValue();
                });
        }

        protected PropertyInstance ImplementId()
        {
            return ImplementPotentialPrimitive(template_p_id, metaData.GetIdMemberByIdIndex((sbyte)idIndex));
        }

        protected PropertyInstance ImplementVersion()
        {
            return ImplementPotentialPrimitive(template_p_version, metaData.VersionMember);
        }

        protected void ImplementToString()
        {
            PropertyInstance p_objRefTemplate = GetObjRefTemplatePI(this);

            MethodInstance methodSb;
		    {
			    methodSb = MethodInstance.FindByTemplate(template_m_toStringSb, true);
			    if (methodSb == null || methodSb.Access.HasFlag(MethodAttributes.Abstract))
			    {
				    IMethodVisitor mg = VisitMethod(template_m_toStringSb);
				    mg.CallThisGetter(p_objRefTemplate);
				    mg.LoadThis();
				    mg.LoadArgs();
				    mg.InvokeVirtual(m_objRef_toStringSb);
				    mg.ReturnValue();
				    mg.EndMethod();
				    methodSb = mg.Method;
			    }
		    }
		    {
			    MethodInstance method = MethodInstance.FindByTemplate(template_m_toString, true);
			    if (method == null || NewType.GetType(typeof(Object)).Equals(method.Owner) || methodSb.Access.HasFlag(MethodAttributes.Abstract))
			    {
				    IMethodVisitor mg = VisitMethod(template_m_toString);
				    LocalVariableInfo loc_sb = mg.NewLocal(typeof(StringBuilder));
				    mg.LoadThis();
                    mg.NewInstance(c_stringBuilder, null);
				    mg.StoreLocal(loc_sb);
				    mg.LoadLocal(loc_sb);
				    mg.InvokeVirtual(methodSb);
				    mg.LoadLocal(loc_sb);
				    mg.InvokeVirtual(new MethodInstance(null, typeof(StringBuilder), typeof(String), "ToString"));
				    mg.ReturnValue();
				    mg.EndMethod();
			    }
		    }
        }

	    protected void ImplementEquals()
	    {
		    PropertyInstance p_objRefTemplate = GetObjRefTemplatePI(this);

		    MethodInstance method = MethodInstance.FindByTemplate(template_m_equals, true);
            if (method == null || method.Access.HasFlag(MethodAttributes.Abstract))
		    {
			    IMethodVisitor mg = VisitMethod(template_m_equals);
			    mg.CallThisGetter(p_objRefTemplate);
			    mg.LoadThis();
			    mg.LoadArgs();
			    mg.InvokeVirtual(m_objRef_equals);
			    mg.ReturnValue();
			    mg.EndMethod();
		    }
	    }

	    protected void ImplementHashCode()
	    {
		    PropertyInstance p_objRefTemplate = GetObjRefTemplatePI(this);

		    MethodInstance method = MethodInstance.FindByTemplate(template_m_hashCode, true);
            if (method == null || method.Access.HasFlag(MethodAttributes.Abstract))
		    {
                IMethodVisitor mg = VisitMethod(template_m_hashCode);
			    mg.CallThisGetter(p_objRefTemplate);
			    mg.LoadThis();
			    mg.LoadArgs();
			    mg.InvokeVirtual(m_objRef_hashCode);
			    mg.ReturnValue();
			    mg.EndMethod();
		    }
	    }
    }
}

File: jambeth/jambeth-merge-bytecode/src/main/java/de/osthus/ambeth/bytecode/visitor/ObjRefVisitor.java
package de.osthus.ambeth.bytecode.visitor;

import java.lang.reflect.Constructor;

import de.osthus.ambeth.bytecode.ClassGenerator;
import de.osthus.ambeth.bytecode.ConstructorInstance;
import de.osthus.ambeth.bytecode.FieldInstance;
import de.osthus.ambeth.bytecode.MethodGenerator;
import de.osthus.ambeth.bytecode.MethodInstance;
import de.osthus.ambeth.bytecode.PropertyInstance;
import de.osthus.ambeth.bytecode.Script;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.merge.model.IEntityMetaData;
import de.osthus.ambeth.merge.model.IObjRef;
import de.osthus.ambeth.metadata.Member;
import de.osthus.ambeth.mixin.ObjRefMixin;
import de.osthus.ambeth.repackaged.org.objectweb.asm.ClassVisitor;
import de.osthus.ambeth.repackaged.org.objectweb.asm.Label;
import de.osthus.ambeth.repackaged.org.objectweb.asm.Opcodes;
import de.osthus.ambeth.repackaged.org.objectweb.asm.Type;
import de.osthus.ambeth.repackaged.org.objectweb.asm.commons.GeneratorAdapter;
import de.osthus.ambeth.util.IConversionHelper;
import de.osthus.ambeth.util.IPrintable;
import de.osthus.ambeth.util.ReflectUtil;

public class ObjRefVisitor extends ClassGenerator
{
	public static final Class<?> templateType = ObjRefMixin.class;

	protected static final String templatePropertyName = templateType.getSimpleName();

	private static final ConstructorInstance c_stringBuilder;

	private static final PropertyInstance template_p_realType = PropertyInstance.findByTemplate(IObjRef.class, "RealType", Class.class, false);

	private static final PropertyInstance template_p_idIndex = PropertyInstance.findByTemplate(IObjRef.class, "IdNameIndex", byte.class, false);

	private static final PropertyInstance template_p_id = PropertyInstance.findByTemplate(IObjRef.class, "Id", Object.class, false);

	private static final PropertyInstance template_p_version = PropertyInstance.findByTemplate(IObjRef.class, "Version", Object.class, false);

	private static final MethodInstance template_m_equals = new MethodInstance(null, Object.class, boolean.class, "equals", Object.class);

	private static final MethodInstance template_m_hashCode = new MethodInstance(null, Object.class, int.class, "hashCode");

	private static final MethodInstance template_m_toString = new MethodInstance(null, Object.class, String.class, "toString");

	private static final MethodInstance template_m_toStringSb = new MethodInstance(null, IPrintable.class, void.class, "toString", StringBuilder.class);

	private static final MethodInstance m_objRef_equals = new MethodInstance(null, templateType, boolean.class, "objRefEquals", IObjRef.class, Object.class);

	private static final MethodInstance m_objRef_hashCode = new MethodInstance(null, templateType, int.class, "objRefHashCode", IObjRef.class);

	private static final MethodInstance m_objRef_toStringSb = new MethodInstance(null, templateType, void.class, "objRefToString", IObjRef.class,
			StringBuilder.class);

	static
	{
		try
		{
			c_stringBuilder = new ConstructorInstance(StringBuilder.class.getDeclaredConstructor());
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	public static PropertyInstance getObjRefTemplatePI(ClassGenerator cv)
	{
		Object bean = getState().getBeanContext().getService(templateType);
		PropertyInstance pi = getState().getProperty(templatePropertyName, bean.getClass());
		if (pi != null)
		{
			return pi;
		}
		return cv.implementAssignedReadonlyProperty(templatePropertyName, bean);
	}

	public static PropertyInstance getConversionHelperPI(ClassGenerator cv)
	{
		Object bean = getState().getBeanContext().getService(IConversionHelper.class);
		PropertyInstance pi = getState().getProperty("ConversionHelper", bean.getClass());
		if (pi != null)
		{
			return pi;
		}
		return cv.implementAssignedReadonlyProperty("ConversionHelper", bean);
	}

	protected final IEntityMetaData metaData;

	protected final int idIndex;

	public ObjRefVisitor(ClassVisitor cv, IEntityMetaData metaData, int idIndex)
	{
		super(new InterfaceAdder(cv, IObjRef.class));
		this.metaData = metaData;
		this.idIndex = idIndex;
	}

	@Override
	public void visitEnd()
	{
		implementRealType();
		implementIdIndex();
		PropertyInstance p_id = implementId();
		PropertyInstance p_version = implementVersion();
		implementDefaultConstructor();
		implementIdVersionConstructor(p_id, p_version);
		implementToString();
		implementEquals();
		implementHashCode();

		super.visitEnd();
	}

	protected void implementDefaultConstructor()
	{
		MethodGenerator mv = visitMethod(ConstructorInstance.defaultConstructor);
		mv.loadThis();
		mv.invokeSuperOfCurrentMethod();
		mv.returnValue();
		mv.endMethod();
	}

	protected void implementIdVersionConstructor(final PropertyInstance p_id, final PropertyInstance p_version)
	{
		ConstructorInstance ci_super;
		try
		{
			Constructor<?> superConstructor = getState().getCurrentType().getDeclaredConstructor();
			ci_super = new ConstructorInstance(superConstructor);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
		MethodGenerator mv = visitMethod(new ConstructorInstance(Opcodes.ACC_PUBLIC, null, Object.class, Object.class));
		mv.loadThis();
		mv.invokeConstructor(ci_super);
		mv.callThisSetter(p_id, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.loadArg(0);
			}
		});
		mv.callThisSetter(p_version, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.loadArg(1);
			}
		});
		mv.returnValue();
		mv.endMethod();
	}

	protected void implementRealType()
	{
		implementProperty(template_p_realType, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.push(metaData.getEntityType());
				mg.returnValue();
			}
		}, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.throwException(Type.getType(UnsupportedOperationException.class), "Property is read-only");
				mg.returnValue();
			}
		});
	}

	protected void implementIdIndex()
	{

		implementProperty(template_p_idIndex, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.push(idIndex);
				mg.returnValue();
			}
		}, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.throwException(Type.getType(UnsupportedOperationException.class), "Property is read-only");
				mg.returnValue();
			}
		});
	}

	protected PropertyInstance implementPotentialPrimitive(PropertyInstance property, final Member member)
	{
		if (member == null)
		{
			return implementProperty(property, new Script()
			{
				@Override
				public void execute(MethodGenerator mg)
				{
					mg.pushNull();
					mg.returnValue();
				}
			}, new Script()
			{
				@Override
				public void execute(MethodGenerator mg)
				{
					Label l_isNull = mg.newLabel();

					mg.loadArg(0);
					mg.ifNull(l_isNull);
					mg.throwException(Type.getType(UnsupportedOperationException.class), "Property is read-only");
					mg.mark(l_isNull);
					mg.returnValue();
				}
			});
		}
		final PropertyInstance p_conversionHelper = getConversionHelperPI(this);

		final MethodInstance m_convertValueToType = new MethodInstance(ReflectUtil.getDeclaredMethod(false, IConversionHelper.class, Object.class,
				"convertValueToType", Class.class, Object.class));

		final Type type = Type.getType(member.getRealType());
		final FieldInstance field = implementField(new FieldInstance(Opcodes.ACC_PRIVATE, "f_" + property.getName(), null, type));
		return implementProperty(property, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				if (member.getRealType().isPrimitive())
				{
					Label l_isNull = mg.newLabel();
					int loc_value = mg.newLocal(field.getType());

					mg.getThisField(field);
					mg.storeLocal(loc_value);

					mg.loadLocal(loc_value);
					mg.ifZCmp(field.getType(), GeneratorAdapter.EQ, l_isNull);

					mg.loadLocal(loc_value);
					mg.box(type);
					mg.returnValue();

					mg.mark(l_isNull);
					mg.pushNull();
					mg.returnValue();
				}
				else
				{
					mg.getThisField(field);
					mg.returnValue();
				}
			}
		}, new Script()
		{
			@Override
			public void execute(MethodGenerator mg)
			{
				mg.putThisField(field, new Script()
				{
					@Override
					public void execute(MethodGenerator mg)
					{
						Label l_isNull = mg.newLabel();
						Label l_finish = mg.newLabel();

						mg.loadArg(0);
						mg.ifNull(l_isNull);

						mg.callThisGetter(p_conversionHelper);
						mg.push(type);
						mg.loadArg(0);
						mg.invokeVirtual(m_convertValueToType);

						mg.unbox(type);
						mg.goTo(l_finish);

						mg.mark(l_isNull);
						mg.pushNullOrZero(field.getType());

						mg.mark(l_finish);
					}
				});
				mg.returnValue();
			}
		});
	}

	protected PropertyInstance implementId()
	{
		return implementPotentialPrimitive(template_p_id, metaData.getIdMemberByIdIndex((byte) idIndex));
	}

	protected PropertyInstance implementVersion()
	{
		return implementPotentialPrimitive(template_p_version, metaData.getVersionMember());
	}

	protected void implementToString()
	{
		PropertyInstance p_objRefTemplate = getObjRefTemplatePI(this);

		MethodInstance methodSb;
		{
			methodSb = MethodInstance.findByTemplate(template_m_toStringSb, true);
			if (methodSb == null || (methodSb.getAccess() & Opcodes.ACC_ABSTRACT) != 0)
			{
				MethodGenerator mg = visitMethod(template_m_toStringSb);
				mg.callThisGetter(p_objRefTemplate);
				mg.loadThis();
				mg.loadArgs();
				mg.invokeVirtual(m_objRef_toStringSb);
				mg.returnValue();
				mg.endMethod();
				methodSb = mg.getMethod();
			}
		}
		{
			MethodInstance method = MethodInstance.findByTemplate(template_m_toString, true);
			if (method == null || Type.getType(Object.class).equals(method.getOwner()) || (method.getAccess() & Opcodes.ACC_ABSTRACT) != 0)
			{
				MethodGenerator mg = visitMethod(template_m_toString);
				int loc_sb = mg.newLocal(StringBuilder.class);
				mg.loadThis();
				mg.newInstance(c_stringBuilder, null);
				mg.storeLocal(loc_sb);
				mg.loadLocal(loc_sb);
				mg.invokeVirtual(methodSb);
				mg.loadLocal(loc_sb);
				mg.invokeVirtual(new MethodInstance(null, StringBuilder.class, String.class, "toString"));
				mg.returnValue();
				mg.endMethod();
			}
		}
	}

	protected void implementEquals()
	{
		PropertyInstance p_objRefTemplate = getObjRefTemplatePI(this);

		MethodInstance method = MethodInstance.findByTemplate(template_m_equals, true);
		if (method == null || (method.getAccess() & Opcodes.ACC_ABSTRACT) != 0)
		{
			MethodGenerator mg = visitMethod(template_m_equals);
			mg.callThisGetter(p_objRefTemplate);
			mg.loadThis();
			mg.loadArgs();
			mg.invokeVirtual(m_objRef_equals);
			mg.returnValue();
			mg.endMethod();
		}
	}

	protected void implementHashCode()
	{
		PropertyInstance p_objRefTemplate = getObjRefTemplatePI(this);

		MethodInstance method = MethodInstance.findByTemplate(template_m_hashCode, true);
		if (method == null || (method.getAccess() & Opcodes.ACC_ABSTRACT) != 0)
		{
			MethodGenerator mg = visitMethod(template_m_hashCode);
			mg.callThisGetter(p_objRefTemplate);
			mg.loadThis();
			mg.loadArgs();
			mg.invokeVirtual(m_objRef_hashCode);
			mg.returnValue();
			mg.endMethod();
		}
	}
}
